File: git/py/mygit/tests/test_tree.py
# -*- coding: utf-8 -*-
"""tree 의 시험 — SPEC.md §4.3 · §8.2, 4단계 "정렬 규칙이 전부다".

오라클은 golden/trees/ — 경우마다 진짜 git 이 인덱스에 올린
(모드·blob·경로) 목록(<경우>.tsv)과 write-tree 의 이름(trees.tsv),
그리고 ls-tree -r -t 의 출력(<경우>.ls)이다.
"""
import os
import shutil
import tempfile
import unittest

from mygit import cli, objects, tree, worktree, zlib
from mygit.tests import golden

CASES = golden.tsv('trees', 'trees.tsv')
OBJS = golden.tsv('objects', 'objects.tsv')


def entries_of(case):
    """<경우>.tsv → [(모드, blob 이름, 경로 바이트)]."""
    out = []
    for line in golden.read('trees', case + '.tsv').split(b'\n'):
        if not line or line.startswith(b'#'):
            continue
        mode, oid, path = line.split(b'\t', 2)
        out.append((mode.decode(), oid.decode(), path))
    return out


def unquote(p):
    """ls-tree 가 따옴표로 감싼 경로를 바이트로 되돌린다(시험 전용)."""
    if not p.startswith('"'):
        return p.encode()
    return golden.unescape_c(p[1:-1])


class Git(unittest.TestCase):
    def setUp(self):
        self.tmp = tempfile.mkdtemp()
        self.gitdir = os.path.join(self.tmp, '.git')
        os.makedirs(os.path.join(self.gitdir, 'objects', 'pack'))

    def tearDown(self):
        for root, _d, files in os.walk(self.tmp):
            for f in files:
                os.chmod(os.path.join(root, f), 0o644)
        shutil.rmtree(self.tmp)


class TestWriteTree(Git):
    def test_s4_3_twelve_trees_have_git_names(self):
        self.assertEqual(len(CASES), 12)
        for row in CASES:
            got = tree.write_tree(self.gitdir, entries_of(row['case']))
            self.assertEqual(got, row['tree'], row['case'])

    def test_s4_3_every_subtree_git_listed_was_written(self):
        for row in CASES:
            tree.write_tree(self.gitdir, entries_of(row['case']))
            ls = golden.read('trees', row['case'] + '.ls').decode()
            for line in ls.split('\n'):
                if not line:
                    continue
                meta = line.split('\t')[0].split(' ')
                if meta[1] == 'tree':
                    t, _ = objects.read_object(self.gitdir, meta[2])
                    self.assertEqual(t, 'tree', line)

    def test_s4_3_flatten_gives_back_the_input(self):
        for row in CASES:
            ents = entries_of(row['case'])
            oid = tree.write_tree(self.gitdir, ents)
            flat = tree.flatten_tree(self.gitdir, oid)
            self.assertEqual(flat, ents, row['case'])


class TestSortRule(unittest.TestCase):
    def test_s4_3_directory_sorts_as_if_it_ended_in_slash(self):
        z = '0' * 40
        ents = [('100644', b'ab', z), ('40000', b'a', z),
                ('100644', b'a=b', z), ('100644', b'a.b', z),
                ('100644', b'a-b', z)]
        body = tree.serialize_tree(ents)
        names = [n for _m, n, _o in tree.parse_tree(body)]
        self.assertEqual(names, [b'a-b', b'a.b', b'a', b'a=b', b'ab'])

    def test_s4_3_plain_name_sort_would_be_wrong(self):
        # 이름만으로 정렬하면 a 가 맨 앞 — 규칙이 왜 있는지 보여 준다
        self.assertLess(tree.tree_entry_key('100644', b'a-b'),
                        tree.tree_entry_key('40000', b'a'))
        self.assertLess(tree.tree_entry_key('100644', b'a'),
                        tree.tree_entry_key('100644', b'a-b'))

    def test_s4_3_mode_is_not_zero_padded_in_the_body(self):
        body = tree.serialize_tree([('40000', b'd', '1' * 40)])
        self.assertTrue(body.startswith(b'40000 d\0'))


class TestParse(unittest.TestCase):
    def test_s4_3_round_trip_on_git_trees(self):
        n = 0
        for row in OBJS:
            if row['type'] != 'tree':
                continue
            raw = zlib.decompress(golden.read('objects', row['id']))
            body = raw.partition(b'\0')[2]
            self.assertEqual(tree.serialize_tree(tree.parse_tree(body)),
                             body)
            n += 1
        self.assertGreaterEqual(n, 2)


class TestQuote(unittest.TestCase):
    def test_s8_2_quoting(self):
        q = worktree.quote_path
        self.assertEqual(q(b'plain.txt'), 'plain.txt')
        self.assertEqual(q(b'sp ace'), 'sp ace')
        self.assertEqual(q(b'sp ace', space=True), '"sp ace"')
        self.assertEqual(q(b'tab\tx'), '"tab\\tx"')
        self.assertEqual(q(b'q"uote'), '"q\\"uote"')
        self.assertEqual(q(b'back\\slash'), '"back\\\\slash"')
        self.assertEqual(q('한글.txt'.encode()),
                         '"\\355\\225\\234\\352\\270\\200.txt"')
        self.assertEqual(q(b'del\x7f'), '"del\\177"')


class TestCatFileTree(Git):
    def test_s9_cat_file_p_tree_matches_ls_tree_top_level(self):
        root = os.path.dirname(self.gitdir)
        env = dict(os.environ, GIT_CEILING_DIRECTORIES=self.tmp)
        for row in CASES:
            tree.write_tree(self.gitdir, entries_of(row['case']))
            ls = golden.read('trees', row['case'] + '.ls').decode()
            want = ''.join(line + '\n' for line in ls.split('\n')
                           if line and b'/' not in
                           unquote(line.split('\t', 1)[1]))
            code, out, _ = cli.run(['cat-file', '-p', row['tree']],
                                   cwd=root, env=env)
            self.assertEqual((code, out.decode()), (0, want),
                             row['case'])


if __name__ == '__main__':
    unittest.main()

File: git/cpp/tests/test_tree.cpp
// tree 의 시험 — SPEC.md §4.3 · §8.2, 4단계 "정렬 규칙이 전부다".
// 오라클은 golden/trees/ — 경우마다 진짜 git 이 인덱스에 올린 목록
// (<경우>.tsv), write-tree 의 이름(trees.tsv), ls-tree -r -t 의 출력
// (<경우>.ls).
#include "golden.hpp"

using namespace mygit;

namespace {
std::vector<PathEntry> entries_of(const std::string& c) {
    std::vector<PathEntry> out;
    for (auto& line :
         golden::split(golden::read("trees/" + c + ".tsv"), '\n')) {
        if (line.empty() || line[0] == '#') continue;
        auto t1 = line.find('\t'), t2 = line.find('\t', t1 + 1);
        out.push_back({line.substr(0, t1),
                       line.substr(t1 + 1, t2 - t1 - 1),
                       line.substr(t2 + 1)});
    }
    return out;
}

// unquote 는 git 이 C 식으로 감싼 경로를 바이트로 되돌린다(시험 전용).
std::string unquote(const std::string& p) {
    if (p.empty() || p[0] != '"') return p;
    std::string s = p.substr(1, p.size() - 2), out;
    const std::string esc = "abtnvfr\"\\";
    const char val[] = {7, 8, 9, 10, 11, 12, 13, '"', '\\'};
    for (size_t i = 0; i < s.size(); ++i) {
        if (s[i] != '\\') {
            out += s[i];
        } else if (auto k = esc.find(s[i + 1]); k != esc.npos) {
            out += val[k], ++i;
        } else {
            out += char(std::stoi(s.substr(i + 1, 3), nullptr, 8));
            i += 3;
        }
    }
    return out;
}
}  // namespace

TEST(s4_3_twelve_trees_have_git_names) {
    auto cases = golden::tsv("trees/trees.tsv");
    CHECK_EQ(cases.size(), size_t(12));
    for (auto& r : cases) {
        golden::TempDir t;
        auto g = golden::gitdir(t);
        auto ents = entries_of(r["case"]);
        auto oid = write_tree(g, ents);
        CHECK_EQ(oid, r["tree"]);
        for (auto& line : golden::split(
                 golden::read("trees/" + r["case"] + ".ls"), '\n')) {
            auto meta =
                golden::split(line.substr(0, line.find('\t')), ' ');
            if (meta.size() == 3 && meta[1] == "tree")
                CHECK_EQ(read_object(g, meta[2]).type, "tree");
        }
        CHECK(flatten_tree(g, oid) == ents);
    }
}

TEST(s4_3_directory_sorts_as_if_it_ended_in_slash) {
    std::string z(40, '0');
    auto body = serialize_tree({{"100644", "ab", z},
                                {"40000", "a", z},
                                {"100644", "a=b", z},
                                {"100644", "a.b", z},
                                {"100644", "a-b", z}});
    std::vector<std::string> names;
    for (auto& e : parse_tree(body)) names.push_back(e.name);
    CHECK((names ==
           std::vector<std::string>{"a-b", "a.b", "a", "a=b", "ab"}));
    // 이름만으로 정렬하면 a 가 맨 앞 — 규칙이 왜 있는지 보여 준다
    CHECK(tree_entry_key("100644", "a-b") < tree_entry_key(DIR, "a"));
    CHECK(tree_entry_key("100644", "a") <
          tree_entry_key("100644", "a-b"));
    CHECK(serialize_tree({{DIR, "d", std::string(40, '1')}})
              .starts_with(std::string("40000 d\0", 8)));
    golden::git_error(
        [] { parse_tree(std::string("100644 x\0abc", 12)); });
}

TEST(s4_3_round_trip_on_git_trees) {
    int n = 0;
    for (auto& r : golden::tsv("objects/objects.tsv")) {
        if (r["type"] != "tree") continue;
        auto raw = decompress(golden::read("objects/" + r["id"]));
        auto body = raw.substr(raw.find('\0') + 1);
        CHECK(serialize_tree(parse_tree(body)) == body);
        ++n;
    }
    CHECK(n >= 2);
}

TEST(s8_2_quoting) {
    CHECK_EQ(quote_path("plain.txt"), "plain.txt");
    CHECK_EQ(quote_path("sp ace"), "sp ace");
    CHECK_EQ(quote_path("sp ace", true), "\"sp ace\"");
    CHECK_EQ(quote_path("tab\tx"), "\"tab\\tx\"");
    CHECK_EQ(quote_path("q\"uote"), "\"q\\\"uote\"");
    CHECK_EQ(quote_path("back\\slash"), "\"back\\\\slash\"");
    CHECK_EQ(quote_path("한글.txt"),
             "\"\\355\\225\\234\\352\\270\\200.txt\"");
    CHECK_EQ(quote_path("del\x7f"), "\"del\\177\"");
}

TEST(s9_cat_file_p_tree_matches_ls_tree) {
    for (auto& r : golden::tsv("trees/trees.tsv")) {
        golden::Sandbox s(true);
        write_tree(s.root + "/.git", entries_of(r["case"]));
        std::string want;
        for (auto& line : golden::split(
                 golden::read("trees/" + r["case"] + ".ls"), '\n')) {
            auto tab = line.find('\t');
            if (tab != line.npos &&
                unquote(line.substr(tab + 1)).find('/') ==
                    std::string::npos)
                want += line + "\n";
        }
        auto res = s.mygit({"cat-file", "-p", r["tree"]});
        CHECK_EQ(res.code, 0);
        CHECK_EQ(res.out, want);
    }
}
